frontend/test/lit/test_quantum_control.py

```diff
@@ -0,0 +1,163 @@
+        ):
+            self.backend_name = backend_name if backend_name else "default"
+            self.backend_lib = backend_lib if backend_lib else "default"
+            self.backend_kwargs = backend_kwargs if backend_kwargs else ""
+            super().__init__(wires=wires, shots=shots)
+
+        def apply(self, operations, **kwargs):  # pylint: disable=missing-function-docstring
+            pass
+
+        @staticmethod
+        def get_c_interface():
+            """Location to shared object with C/C++ implementation"""
+            return get_lib_path("runtime", "RUNTIME_LIB_DIR") + "/libdummy_device.so"
+
+    return CustomDevice(wires=num_wires)
+
+
+def test_named_controlled():
+    """Test that named-controlled operations are passed as-is."""
+    dev = get_custom_device(2, set(), set())
+
+    @qjit(target="mlir")
+    @qfunc(device=dev)
+    # CHECK-LABEL: public @jit_named_controlled
+    def named_controlled():
+        # CHECK: quantum.custom "CNOT"
+        qml.CNOT(wires=[0, 1])
+        # CHECK: quantum.custom "CY"
+        qml.CY(wires=[0, 1])
+        # CHECK: quantum.custom "CZ"
+        qml.CZ(wires=[0, 1])
+        return measure(wires=0)
+
+    print(named_controlled.mlir)
+
+
+test_named_controlled()
+
+
+def test_native_controlled_custom():
+    """Test native control of a custom operation."""
+    dev = get_custom_device(3, discarded_operations={"CRot"}, added_operations={"Rot", "C(Rot)"})
+
+    @qjit(target="mlir")
+    @qfunc(device=dev)
+    # CHECK-LABEL: public @jit_native_controlled
+    def native_controlled():
+        # CHECK: [[out:%.+]], [[out_ctrl:%.+]]:2 = quantum.custom "Rot"
+        # CHECK-SAME: ctrl
+        # CHECK-SAME: ctrlval %true, %true
+        qml.ctrl(qml.Rot(0.3, 0.4, 0.5, wires=[0]), control=[1, 2])
+        return measure(wires=0)
+
+    print(native_controlled.mlir)
+
+
+test_native_controlled_custom()
+
+
+def test_native_controlled_unitary():
+    """Test native control of the unitary operation."""
+    dev = get_custom_device(4, set(), set())
+
+    @qjit(target="mlir")
+    @qfunc(device=dev)
+    # COM: CHECK-LABEL: public @jit_native_controlled_unitary
+    def native_controlled_unitary():
+        # COM: CHECK: [[out:%.+]], [[out_ctrl:%.+]]:3 = quantum.unitary
+        # COM: CHECK-SAME: ctrl
+        # COM: CHECK-SAME: ctrlval %true, %true, %true
+        qml.ctrl(
+            qml.QubitUnitary(
+                jnp.array(
+                    [
+                        [0.70710678 + 0.0j, 0.70710678 + 0.0j],
+                        [0.70710678 + 0.0j, -0.70710678 + 0.0j],
+                    ],
+                    dtype=jnp.complex128,
+                ),
+                wires=[0],
+            ),
+            control=[1, 2, 3],
+        )
+        return measure(wires=0)
+
+    print(native_controlled_unitary.mlir)
+
+
+# TODO: Remove `COM` comments and enable, once the PL fixes the unitary decomposition tracing
+# test_native_controlled_unitary()
```

Comment:
Do you have a ticket opened in PL? Is this behaviour necessary for the release, @dime10?

Reply:
Good question, I would have thought controlled qubit unitary is now supported. What is the issue in PL you are referring to @grwlf ?

Reply:
The development version of PL that I used earlier allowed this test to pass, but our current version did not. Unfortunately I didn't record the exact dev hash. Anyways, now this test started to work after some fixes so I have enabled it and removed the mentioned comments. Thanks for noticing this!